docs/designs/checkpoint_design.md

```diff
@@ -0,0 +1,59 @@
+checkpoint_dir = os.path.join(
+    args.checkpoint_dir, checkpoint_folder_name
+)
+```
+ElasticDL is a Kubernetes-native framework and local directory in the pod will be lost if the pod exits in Kubernetes. So the `checkpoint_dir` should be a persistent directory which will not be lost after the pod exit. Now, we can utilize `hostPath` and `PersistentVolume` to get a persistent directory to save checkpoints in ElasticDL.
+
+**Save model parameters to a checkpoint directory.**
+
+Using AllReduceStrategy in ElasticDL, each worker has all parameters and we can assign any worker to save all parameters to a checkpoint directory. Using ParameterServerStrategy, all model parameters are partitioned on all PS instances. They need to save their owner parameters to the same checkpoint directory at the same step. Saving checkpoints for AllReduceStrategy can be seen as a special case that there is only a PS instance using ParameterServerStrategy. So, we can adopt the same way to save checkpoints for AllReduceStrategy and ParameterServerStrategy.
+
+The parameters on each PS instance may contain non-embedding variables and a part of embedding vectors of each embedding table. We create a `Tensor` protobuf with values for each variable or with values and indices for embedding vectors. To save an embedding table, we also need to create `EmbeddingTableInfo` to save the meta information like `initializer` and `dim`. Then, we create a `Model` protobuf including those `Tensor`s and `EmbeddingTableInfo`s. Finally, we save the `Model` to a  file named `variables-{ps_id}-of-{ps_num}.chkpt` where `ps_id` is the index of a PS instance and `ps_num` is the total number of PS instances. The checkpoint file will be located in a subdirectory `model_v{version}` for each checkpoint version. After the iteration versions of all PS instances exceed the checkpoint version in `model_v{version}`,  the `model_v{version}` will contain files of the entire model parameters.
+
+**The maximum number of recent checkpoints to keep**
+
+We can set the `keep_checkpoint_max` in ElasticDL. After saving a checkpoint, each PS instance will check the number of checkpoints it saved and will remove its earliest checkpoint file in `model_v{version}` if the number exceeds the `keep_checkpoint_max`. Then, it will remove the `model_v{version}` subdirectory if it is empty.
+
+### Restore Model Parameters from a Checkpoint
+The number of PS instances may be various when we restore model parameters to train or predict. So, we need to reassign variables and embedding vectors in checkpoint to new PS instances. For each new PS instance, it will traverse all protobuf files in the checkpoint directory and get variables and embedding vectors if `hash_utils.string_to_id(var.name)` or `hash_utils.int_to_id(embedding_id)` is equal to its index (`ps_id`). The pseudo-code is:
+```python
+non_embedding_vars = {}
+embedding_table_vectors = {}
+for pb_file in os.listdir(checkpoint_dir):
```

Comment:
The time complexity is too high. We could just traverse the checkpoint just one time, and restore all PS instances.

Reply:
If we only traverse the checkpoint on a single instance just one time, we need to send partitioned variables from the instance to all PS instances by GRPC. It may be more complex than each PS instance traverses the checkpoint and load its owner parameters.

Reply:
Can you have suggestions to traverse the checkpoint on a single instance just one time?

Reply:
I mean that we could get params for all PS instances in one time traversal the checkpoint.

```python
ps_params = {}
for pb_file in os.listdir(checkpoint_dir):
    model_pb = load_model_file(pb_file)
    for param in model_pb.params:
        if param.indices
            embedding_table_vectors.setdefault(param.name, ([], []))
            for embedding_id, vector in zip(param.indices, param.values):
                ps_id = hash_util.int_to_id(embedding_id)
                if ps_id not in ps_params:
                    ps_params[ps_id] = {"non_embedding" : {}, "embedding": {}}
                if param.name not in ps_params[ps_id]["embedding"]:
                    ps_params[ps_id]["embedding"] = ([embedding_id], [vector])
                ps_params[ps_id]["embedding][param.name][0].append(embedding_id)
                ps_params[ps_id]["embedding][param.name][1].append(embedding_id)
        else:
            ps_id = hash_utils.string_to_id(param.name)
            if ps_id not in ps_vars:
                ps_params[ps_id] = {"non_embedding" : {}, "embedding": {}}
            ps_params[ps_id]["non_embedding"][param.name] = param.values
```

Then, we could send params in `ps_params` one PS by one PS.

Reply:
> I mean that we could get params for all PS instances in one time traversal the checkpoint.
> 
> ```python
> ps_params = {}
> for pb_file in os.listdir(checkpoint_dir):
>     model_pb = load_model_file(pb_file)
>     for param in model_pb.params:
>         if param.indices
>             embedding_table_vectors.setdefault(param.name, ([], []))
>             for embedding_id, vector in zip(param.indices, param.values):
>                 ps_id = hash_util.int_to_id(embedding_id)
>                 if ps_id not in ps_params:
>                     ps_params[ps_id] = {"non_embedding" : {}, "embedding": {}}
>                 if param.name not in ps_params[ps_id]["embedding"]:
>                     ps_params[ps_id]["embedding"][param.name] = ([embedding_id], [vector])
>                 ps_params[ps_id]["embedding][param.name][0].append(embedding_id)
>                 ps_params[ps_id]["embedding][param.name][1].append(embedding_id)
>         else:
>             ps_id = hash_utils.string_to_id(param.name)
>             if ps_id not in ps_vars:
>                 ps_params[ps_id] = {"non_embedding" : {}, "embedding": {}}
>             ps_params[ps_id]["non_embedding"][param.name] = param.values
> ```
> 
> Then, we could send params in `ps_params` one PS by one PS.

I think you means one pod load all the shards of the checkpoint and send to all the PS instance. The memory cost may be high.